Add support for simulators running on a different host than the PX4 instance

FernandezR opened the pull request:
Please use [PX4 Discuss](http://discuss.px4.io/) or [Slack](http://slack.px4.io/) to align on pull requests if necessary. You can then open draft pull requests to get early feedback.

**Describe problem solved by this pull request**
Allows for the use of simulators running on a different host than the PX4 instance.

**Describe your solution**
Added use of `PX4_SIM_HOSTNAME` and `PX4_SIM_HOST_ADDR` environment variables to pass simulator hostname or ip address through the posix rcs script to the simulator module.

**Describe possible alternatives**
Didn't consider any other alternatives as this was the simplest implementation.

**Test data / coverage**
Tested using a docker compose implementation using a a private Unity Simulation, where simulator and PX4 instances exist in different containers. Also, tested on local machine to ensure continued support of `localhost` simulator.

**Additional context**
None

Jaeyoung-Lim commented on ROMFS/px4fmu_common/init.d-posix/rcS:
```diff
@@ -207,7 +207,13 @@ dataman start
+    simulator start -c $simulator_tcp_port
+  else
+    echo "PX4 SIM HOST: $PX4_SIM_HOST"
+    simulator start -c $simulator_tcp_port $PX4_SIM_HOST
```
Wouldn't it make more sense if the host is passed through a flag as it is done with the ports?

FernandezR replied:
I didn't use a flag since currently the flag is only used to specify the protocol. I decided to use the length of the arguments as it was already being used to verify that the port was passed through.

Jaeyoung-Lim commented on src/modules/simulator/simulator_mavlink.cpp:
```diff
@@ -682,9 +684,22 @@ void Simulator::run()
-	_myaddr.sin_addr.s_addr = htonl(INADDR_ANY);
 	_myaddr.sin_port = htons(_port);
 
+	if (_host == "") {
```
```suggestion
	if (_host.empty()) {
```

Jaeyoung-Lim commented on src/modules/simulator/simulator.cpp:
```diff
@@ -95,6 +107,8 @@ static void usage()
 	PX4_INFO("Start simulator:     simulator start");
 	PX4_INFO("Connect using UDP: simulator start -u udp_port");
 	PX4_INFO("Connect using TCP: simulator start -c tcp_port");
+	PX4_INFO("Connect using UDP: simulator start -u udp_port hostname");
```
Is this use case valid / tested?

FernandezR replied:
The UDP version hasn't been tested, I don't have a simulator at this time that uses a UDP connection.

Jaeyoung-Lim replied:
If it is not tested, would it be possible to disable it?

FernandezR replied:
Yes, I can remove it for now or comment it out. Which is preferred?

lovettchris replied:
Can you also make it so that you support passing an explicit ip address instead of hostname ?  In some cases (like windows subsystem for linux WSL2) the host name is the same for the linux and windows machines, but the ip addresses are different.  Explicit ip address can also be handy to select a specific ethernet interface which can be handy when there's multiple ways to reach the host (e.g. wifi versus hard wired).

Jaeyoung-Lim reviewed (commented):
Thanks for the contribution! 

I was curious if the testing done also with the UDP connection? and a few minor comments in line

FernandezR commented:
What commit should I rebase the PR against? I was working off the latest commit in master. The two test that were failing were complaining about not being able to connect to a display.

Jaeyoung-Lim commented:
> The two test that were failing were complaining about not being able to connect to a display.

I don't think this is accurate since it is reliably passing on other PRs - can you hvae a look?

FernandezR commented:
I added the couple of commits that occurred since I made this PR. I'm waiting to see if all the tests pass now.

FernandezR commented:
Two tests are still failing but it is not caused by the changes I made.

- The MacOS build test is failing due to a failed python import of the toml library.
- The SITL Test is failing due to this error: `[gzserver    ] [Err] [RenderEngine.cc:744] Can't open display: :99`

Jaeyoung-Lim commented:
@FernandezR I am pretty sure that the following error is not causing the SITL test to fail, if you look at other PRs the tests pass even though it displays the same message
```
[gzserver    ] [Err] [RenderEngine.cc:744] Can't open display: :99
```

FernandezR commented:
@Jaeyoung-Lim  The specific tests that are not passing are these:
- standard_vtol:
     - 'Fly forward in position control': failed
     - 'Fly forward in altitude control': cancelled

I'm not sure how I'm supposed to fix these tests failing as nothing I changed is impacting them. Looking in the logs the simulator connection is established on localhost as expected.

And I can't do anything to solve the MacOS test that is failing as it has something to do with the CI environment as far as I can tell.

FernandezR commented:
This seems to be the cause of the failure:

`[mavsdk_tests] ../../../test/mavsdk_tests/autopilot_tester.cpp:106: FAILED:
[mavsdk_tests]   CHECK( ground_truth_horizontal_position_far_from(_home, min_distance_m) )
[mavsdk_tests] with expansion:
[mavsdk_tests]   false`

FernandezR commented:
The latest commit on master seems to be failing the same SITL test

FernandezR commented:
@Jaeyoung-Lim Do you have any suggestions of how to debug the failing test? I'm not sure why this assertion, `CHECK( ground_truth_horizontal_position_far_from(_home, min_distance_m) )`,would be failing.

FernandezR commented:
@Jaeyoung-Lim It looks like all the checks finally passed.

Jaeyoung-Lim reviewed (dismissed):
Thanks for looking into the failures, and sorry for the late response.

Looks good to me on my side! Thanks!

dayjaby commented:
@FernandezR can you align this PR with the changes made in https://github.com/PX4/PX4-Autopilot/pull/15443? I like your addition of an environment variable to change the IP.

FernandezR commented:
> @FernandezR can you align this PR with the changes made in #15443? I like your addition of an environment variable to change the IP.

Yes, I will work on aligning with change to support both ip address and hostname for remote simulator.

FernandezR commented:
I updated the request to align to PR  #15443

FernandezR commented:
Failed tests are failing due to this error.

'Run ./Tools/upload_log.py -q --description "${GITHUB_WORKFLOW} ${GITHUB_RUN_ID}" --feedback "${GITHUB_WORKFLOW} ${GITHUB_RUN_ID} ${GITHUB_REPOSITORY} ${GITHUB_REF}" --source CI ~/.ros/log/*/*.ulg
Traceback (most recent call last):
  File "./Tools/upload_log.py", line 127, in <module>
    main()
  File "./Tools/upload_log.py", line 115, in main
    with open(file_name, 'rb') as f:
FileNotFoundError: [Errno 2] No such file or directory: '/github/home/.ros/log/*/*.ulg'
Error: Process completed with exit code 1.`

mrpollo commented:
@Jaeyoung-Lim this PR looks like it would be super useful, and I think we should take over from @FernandezR, could you please branch out, fix the conflicts, and make a new PR? It would be the best way to go since the author already went out of his way to accommodate all of our requests.

FernandezR commented:
> @Jaeyoung-Lim this PR looks like it would be super useful, and I think we should take over from @FernandezR, could you please branch out, fix the conflicts, and make a new PR? It would be the best way to go since the author already went out of his way to accommodate all of our requests.

I have fixed the conflicts and updated to the latest commit in master. Do you still want another pr? This one was updated with the changes I just made automatically.

Also, someone updated the number of arguments being passed to the simulator script but it was incorrect during my testing. So the values are back to there original values.

mrpollo commented:
@FernandezR we don't like merge commits from the main branch, we usually rebase, and to avoid further conflicts with the file you are modifying I have asked @Jaeyoung-Lim to help, he's probably creating a new PR to address the issue. Don't worry you will remain the author of the commits, he's just adding some on top.

lovettchris commented:
I have merged the work Fernandez did with the other change that already added "-t" support for remote ip addresses.  See https://github.com/PX4/PX4-Autopilot/pull/17275

FernandezR commented:
@lovettchris Thanks for merging in my changes. Should we close this pull request now?

lovettchris commented:
Yes I think I got all your changes in so if you are happy with what is in the master branch, go ahead and close this one.

FernandezR commented:
Changes merged into master through pull request  #17275.
